src/core/server/integration_tests/http/lifecycle_handlers.test.ts

```diff
@@ -247,4 +285,102 @@ describe('core lifecycle handlers', () => {
+        (context, req, res) => {
+          return res.ok({ body: 'ok()' });
+        }
+      );
+      router.get(
+        { path: testPublicRoute, validate: false, options: { access: 'public' } },
+        (context, req, res) => {
+          return res.ok({ body: 'ok()' });
+        }
+      );
+      await server.start();
+    });
+
+    it('accepts requests with the internal product header to internal routes', async () => {
+      await supertest(innerServer.listener)
+        .get(testInternalRoute)
+        .set(internalProductHeader, 'anything')
+        .expect(200, 'ok()');
+    });
+
+    it('accepts requests with the internal product header to public routes', async () => {
+      await supertest(innerServer.listener)
+        .get(testPublicRoute)
+        .set(internalProductHeader, 'anything')
+        .expect(200, 'ok()');
+    });
+  });
+});
+
+describe('core lifecyle handers with restrict internal routes enforced', () => {
+  let server: HttpService;
+  let innerServer: HttpServerSetup['server'];
+  let router: IRouter;
+
+  beforeEach(async () => {
+    const configService = configServiceMock.create();
+    configService.atPath.mockImplementation((path) => {
+      if (path === 'server') {
+        return new BehaviorSubject(setUpDefaultServerConfig({ enabled: true }));
+      }
+      if (path === 'externalUrl') {
+        return new BehaviorSubject({
+          policy: [],
+        } as any);
+      }
+      if (path === 'csp') {
+        return new BehaviorSubject({
+          strict: false,
+          disableEmbedding: false,
+          warnLegacyBrowsers: true,
+        });
+      }
+
+      throw new Error(`Unexpected config path: ${path}`);
+    });
+    server = createHttpServer({ configService });
+
+    await server.preboot({ context: contextServiceMock.createPrebootContract() });
+    const serverSetup = await server.setup(setupDeps);
+    router = serverSetup.createRouter('/');
+    innerServer = serverSetup.server;
+  });
+
+  afterEach(async () => {
+    await server.stop();
+  });
+
+  describe('restrictInternalRoutes postauth handler', () => {
+    const testInternalRoute = '/restrict_internal_routes/test/route_internal';
+    const testPublicRoute = '/restrict_internal_routes/test/route_public';
+    beforeEach(async () => {
+      router.get(
+        { path: testInternalRoute, validate: false, options: { access: 'internal' } },
+        (context, req, res) => {
+          return res.ok({ body: 'ok()' });
+        }
+      );
+      router.get(
+        { path: testPublicRoute, validate: false, options: { access: 'public' } },
+        (context, req, res) => {
+          return res.ok({ body: 'ok()' });
+        }
+      );
+      await server.start();
+    });
+
+    it('request requests without the internal product header to internal routes', async () => {
+      const result = await supertest(innerServer.listener).get(testInternalRoute).expect(400);
```

Comment:
Can we add a test case for the request succeeding too?

Reply:
yip, absolutely! Done.